glossary.md

```diff
@@ -0,0 +1,241 @@
+# Glossary
+
+--------------------------------------------------------------------------------
+
+# General Terms
+
+## Layer 1 (L1)
+[L1]: /glossary.md#layer-1-L1
+
+Refers the Ethereum blockchain, used used in contrast to [layer 2][L2], which
+refers to Optimistic Ethereum.
+
+## Layer 2 (L2)
+[L2]: /glossary.md#layer-2-L2
+
+Refers to the Optimistic Ethereum blockchain (specified in this repository),
+used in contrast to [layer 1][L1], which refers to the Ethereum blockchain.
+
+## Block
+[block]: /glossary.md#layer-1-L1
+
+Can refer to an [L1] block, or to an [L2] block, which are structured similarly.
+
+A block is a sequential list of transactions, along with a couple of properties
+stored in the *header* of the block. A description of these properties can be
+found in code comments [here][nano-header], or in the [Ethereum yellow paper
+(pdf)][yellow], section 4.3.
+
+It is useful to distinguish between input block properties, which are known
+before executing the transactions in the block, and output bblock properties,
+which are derived after executing the block's transactions. These include
+various [Merkle roots][Merkle root] for the L2 state and for log events emitted
+during the execution.
+
+## EOA
+[EOA]: /glossary.md#EOA
+
+"Externally Owned Account", an Ethereum term to designate addresses operated by
+users, as opposed to contract addresses.
+
+## Merkle Root
+[Merkle root]: /glossary.md#merkle-roots
+
+The Merkle root is the root hash of a [Merkle Patricia tree] (MPT). A MPT is a
+[trie], which is a tree tree structure that maps keys to values. The root hash
+of a MPT is a commitment to the contents of the tree, which allows a proof to be
+constructed for any key-value mapping encoded in the tree. Such a proof is
+called a Merkle proof, and can be verified against the Merkle root.
+
+## Chain Re-Organization
+[reorg]: /glossary.md#chain-re-organization
+
+A re-organization, or re-org for short, is whenever the head of a blockchain
+(its last block) changes (as dictated by the [fork choice rule]) to a block that
+is not a child of the previous head.
+
+L1 re-orgs can happen because of network conditions or attacks. L2 re-orgs are a
+consequence of L1 re-orgs, mediated via [L2 chain derivation][derivation].
+
+## Precompiled Contract ("Precompile")
+[precompile]: /glossary.md#precompiled-contract-precompile
```

Comment:
Precompile vs predeploy? Does predeploy need to be specified as well?

Reply:
What's the trade-off between these two things? A pre-compile does necessitate a Geth transaction, but maybe it can be done very cleanly?

While we're at it, what do we expect the precompile/predeploys to be besides this L1 attributes oracle?